packages/medusa/src/strategies/product-export.ts

```diff
@@ -0,0 +1,473 @@
+  private readonly BATCH_SIZE = 50
+
+  constructor({
+    manager,
+    batchJobService,
+    productService,
+    fileService,
+    productRepository,
+  }: InjectedDependencies) {
+    super({
+      manager,
+      batchJobService,
+      productService,
+      fileService,
+      productRepository,
+    })
+
+    this.manager_ = manager
+    this.batchJobService_ = batchJobService
+    this.productService_ = productService
+    this.fileService_ = fileService
+    this.productRepository_ = productRepository
+  }
+
+  async buildTemplate(): Promise<string> {
+    return this.buildHeader()
+  }
+
+  async completeJob(batchJobId: string): Promise<BatchJob> {
+    return await this.atomicPhase_(async (transactionManager) => {
+      const batchJob = await this.batchJobService_
+        .withTransaction(transactionManager)
+        .retrieve(batchJobId)
+
+      if (batchJob.status === BatchJobStatus.COMPLETED) {
+        return batchJob
+      }
+
+      return await this.batchJobService_.complete(batchJob)
+    })
+  }
+
+  async prepareBatchJobForProcessing(
+    batchJobId: string,
+    req: Express.Request
+  ): Promise<BatchJob> {
+    return await this.atomicPhase_(async (transactionManager) => {
+      return await this.batchJobService_
+        .withTransaction(transactionManager)
+        .ready(batchJobId /* batchJob after the above todo */)
+    })
+  }
+
+  async processJob(batchJobId: string): Promise<BatchJob> {
+    let offset = 0
+    let advancementCount = 0
+    let productCount = 0
+
+    return await this.atomicPhase_(
+      async (transactionManager) => {
+        let batchJob = await this.batchJobService_
+          .withTransaction(transactionManager)
+          .retrieve(batchJobId)
+
+        offset = batchJob.context.offset ?? offset
+        advancementCount = batchJob.context.count ?? advancementCount
+
+        const {
+          writeStream,
+          url: key,
+          promise,
+        } = await this.fileService_.getUploadStreamDescriptor({
+          name: "product-export",
+          ext: "csv",
+        })
+
+        const header = await this.buildHeader()
+        writeStream.write(header)
+
+        const { listConfig, filterableFields } = batchJob.context as Context
+
+        productCount = await this.productService_
+          .withTransaction(transactionManager)
+          .count(filterableFields)
+
+        while (advancementCount < productCount) {
+          const [products, count] = await this.productService_
+            .withTransaction(transactionManager)
+            .listAndCount(filterableFields, {
+              relations: this.relations_,
+              ...((listConfig as Record<string, unknown>) ?? {}),
+              skip: offset,
+              take: this.BATCH_SIZE,
+              order: { created_at: "DESC" },
+            })
+
+          products.forEach((product) => {
+            writeStream.write(this.buildLine(product))
+          })
+
+          advancementCount += count
```

Comment:
the `count` returned from `productService_.listAndCount` is the total query count without limitations, not how many entities are returned in the query. For maintaining a progress count we should use `offset` and `BATCH_SIZE`

Reply:
Indeed, just read the doc and the count is without limitations 👍 ill do the update

Reply:
I will use products.length since the number of products is not necessarily equal to either offset or batch size when reaching the last page 😃

Reply:
done, let me know what you thing of the last changes 👍